### PR TITLE
Keep Triton FLCE C=1 by default and use C=8 for GPT-OSS

### PR DESCRIPTION
**Pull request by kolehma8**
## Summary

- keep the public Triton fused linear cross entropy chunk-memory default at `C=1`
- add an explicit per-call `chunk_mem_const` override while preserving backward compatibility for existing autograd callers
- use `C=8` only in the GPT-OSS patched training path, where C=1 causes a deterministic BF16 convergence mismatch
- keep every other model and direct/public FLCE caller at C=1
- retain the reusable-logits, direct-dX, and direct-dW optimizations from #1468
- lock the default and override chunk geometry in regression coverage

## Why the override is scoped

The exact mini GPT-OSS convergence case fails with C=1 and passes with C=4, C=8, and C=16. However, setting C=8 globally causes the BF16 Qwen3.5-MoE convergence case to fail. The model-scoped override preserves both trajectories:

- public/default FLCE: C=1
- GPT-OSS patched loss: C=8
- fused linear KL: unchanged independently tuned policy

## H200 memory and latency

BF16, `BT=8192`, `H=4096`, `V=128256`:

| Policy | Chunk size | Chunks | Time | Incremental peak allocation |
|---|---:|---:|---:|---:|
| Public default C=1 | 256 | 32 | 53.039 ms | 1128.66 MiB |
| GPT-OSS override C=8 | 2048 | 4 | 37.748 ms | 1568.04 MiB |
| Previous global C=16 | 4096 | 2 | 35.074 ms | 2068.04 MiB |

The default therefore retains the minimum memory budget, while GPT-OSS uses the smallest requested release override with stable convergence.

## Testing done

On H200 with PyTorch 2.9.1.6+cu129 and Triton 3.6.0:

- **144 passed**: FLCE ops suite
- **144 passed**: FLCE ops suite with `LIGER_KERNEL_STRICT=1`
- **141 passed**: transformer-facing FLCE suite
- **71 passed, 3 dependency-gated skips**: FP32 + BF16 text-model convergence suites
- exact BF16 mini GPT-OSS convergence passes with the scoped C=8 override
- BF16 mini Qwen3.5-MoE convergence passes because its default remains C=1
- Ruff lint/format and `git diff --check` pass

The broader six-file convergence attempt also exposed four PaliGemma multimodal failures in the ad hoc Transformers 5.15 environment. Those configurations explicitly set `fused_linear_cross_entropy=False`, so they are unrelated to this change.

No owned B300 pod was available for an additional Blackwell run.

### CODE REVIEW

**Comment by kolehma8** [comment]
Convergence update: the BF16 mini_gpt_oss case is a real C=1 regression, not flakiness. In the same H200 environment and exact test node, C=1 failed at training step 18 (loss 7.7911 vs 8.3155, outside tolerance), while overriding only `_CHUNK_MEM_CONST` back to C=16 passed. The two FP32 PaliGemma multimodal failures are unrelated because those model setups explicitly use `fused_linear_cross_entropy=False`. Marking this PR draft while the full convergence matrix completes and the GPT-OSS divergence is investigated.

**Review by Mecoli1219 — APPROVED** [review]

**Review by vaibhavjindal — APPROVED** [review]

**Comment by hiwuhgds-pixel** [comment]
I recently re-ran benchmarks to inspect the chunking mechanism in `fused_linear_cross_entropy` and noticed a few noteworthy points regarding the relationship between `BT` and `num_chunks` that I'd like to share for discussion:

Current chunk calculation code:
```python
inc_factor  = triton.cdiv(V, _CHUNK_MEM_CONST * H)
chunk_size  = triton.next_power_of_2(triton.cdiv(BT, inc_factor))
chunk_size  = min(chunk_size, BT)
num_chunks  = triton.cdiv(BT, chunk_size)
```

Here, `inc_factor` only depends on the static parameters `(V, H)` and `_CHUNK_MEM_CONST`, and is completely independent of `BT`. Therefore, for a fixed model architecture, `num_chunks` is virtually pinned (only fluctuating slightly due to power-of-two rounding and `cdiv`). When `BT` changes, in practice only `chunk_size` scales accordingly. In other words, `num_chunks` in this formula only changes when the `V/H` ratio varies.

---

However, empirical benchmark data shows that `chunk_size` (tokens per chunk) plays the primary role in GEMM performance and kernel overhead, rather than the chunk count or the `V/H` ratio.

Specifically, I swept `BT` on a fixed model configuration (V = 128256, H = 4096, benchmarked on B200, BF16, full fwd+bwd):

| BT | num_chunks | tokens/chunk | Runtime (B200, BF16) |
|:---:|:---:|:---:|:---:|
| 8192 | 1 | 8192 | 21.5 ms |
| 8192 | 32 | 256 | 125.4 ms |
| 4096 | 32 | 128 | 120.6 ms |
| 1024 | 32 | 32 | 119.0 ms |

Three points stand out from the data:
- GEMM fragmentation: At the same `BT = 8192`, splitting into 32 chunks causes runtime to surge from 21.5 ms to 125.4 ms (5.82× slower). The M dimension of the GEMM is fragmented from 8192 down to 256, preventing the GPU from fully saturating Tensor Cores, while kernel launch overhead and gradient accumulation costs are repeated 32 times.
- Overhead bottleneck: Keeping `num_chunks = 32` while reducing `BT` from 8192 down to 1024 (an 8× reduction in computational workload) results in almost no change in runtime (~125.4 ms down to 119.0 ms). At this point, each chunk contains only 32 tokens, and kernel execution is dominated entirely by kernel launch overhead and gradient accumulation iterations rather than actual arithmetic computation.
- Chunk overhead tolerance varies significantly with `BT`: At `BT = 2048`, simply moving from 1 to 2 chunks (`chunk_size` reduced to 1024) already causes a 1.435× slowdown; whereas at `BT = 32768`, it takes 16 chunks to reach a 1.484× slowdown, while 2 chunks exhibit a virtually negligible difference (1.018× slowdown because `chunk_size` remains large at 16384).

---

From a memory (VRAM) perspective, the marginal benefit of increasing the chunk count diminishes rapidly:

| Chunks | Buffer/chunk | Reduction vs. Full | Marginal Reduction vs. Previous |
|:---:|:---:|:---:|:---:|
| 1 | 100% | 0% | — |
| 2 | 50% | 50% | 50% |
| 4 | 25% | 75% | 25% |
| 8 | 12.5% | 87.5% | 12.5% |
| 16 | 6.25% | 93.75% | 6.25% |
| 32 | 3.125% | 96.875% | 3.125% |
| 64 | 1.5625% | 98.4375% | 1.5625% |

Going from 16 to 32 chunks saves only an additional 3.125% of logits buffer memory (and going to 64 chunks saves only another ~1.56%). Meanwhile, Peak Loss VRAM always includes `grad_weight`, which has a fixed footprint of `V × H × dtype`. At smaller `BT` values, where the logits buffer already accounts for an insignificant fraction of total Peak VRAM, maintaining a large `num_chunks` barely yields any memory savings, yet incurs a severe runtime penalty.

---

Another key observation: the `V/H` ratio does not correlate with or predict the degree of performance degradation.

I compared two models with a ~35% difference in their `V/H` ratios:
- Llama-3-8B (`V/H ≈ 31.3`)
- Qwen2.5-7B (`V/H ≈ 42.4`)

Comparing the slowdown factor relative to the `num_chunks = 1` baseline on B200 (BF16):

| BT | num_chunks | tokens/chunk | Llama-3-8B (`V/H ≈ 31.3`) | Qwen2.5-7B (`V/H ≈ 42.4`) |
|---:|:---:|:---:|---:|---:|
| 1024 | 2 | 512 | 1.503× | 1.561× |
| 1024 | 4 | 256 | 2.646× | 2.730× |
| 1024 | 8 | 128 | 4.968× | 5.148× |
| 4096 | 2 | 2048 | 1.225× | 1.213× |
| 4096 | 4 | 1024 | 1.761× | 1.701× |
| 4096 | 8 | 512 | 2.803× | 2.722× |
| 32768 | 2 | 16384 | 1.018× | 1.048× |
| 32768 | 4 | 8192 | 1.108× | 1.134× |
| 32768 | 8 | 4096 | 1.207× | 1.228× |

Despite a ~35% difference in `V/H` ratio, the slowdown factors at each `(BT, num_chunks)` point between the two models are remarkably consistent (mostly within a 5% difference). This demonstrates that performance degradation is directly determined by `chunk_size` relative to hardware compute capabilities, rather than the `V/H` ratio.

---

The existing code comment notes that the intention behind `inc_factor = cdiv(V, H)` is *"a memory floor, not a perf target"*, aiming to anchor the logits buffer around `BT × H`. However, in practice, the parameter directly dictating kernel performance is `chunk_size = BT / num_chunks`. The memory-speed trade-off shifts significantly across different `BT` values: VRAM savings experience steep diminishing returns as chunk count increases, while runtime costs increase sharply when `chunk_size` drops below the GPU's efficient GEMM regime and becomes dominated by overhead. Consequently, locking `num_chunks` to the `V/H` ratio results in an uneven and inconsistent trade-off across varying `BT` values.

---

Regarding the case of lowering `C=1` due to convergence requirements:

I swept `num_chunks` from 1 to 32 across `mini GPT-OSS`, `mini Qwen3.5-MoE`, and `mini Qwen3-MoE` (cross-tested on both H200 and B200), and observed that:

- The core difference between `C=1` and `C=8` essentially comes down to chunk count: many chunks (~32) vs. few chunks (~4).
- For GPT-OSS: numerical error accumulates in clear proportion to `num_chunks`. Fewer chunks mean fewer gradient accumulation (`addmm`) steps, resulting in smaller accumulated error and passing tests. In fact, on B200, even `num_chunks = 32` can pass tolerance thresholds that fail on H200.
- For the Qwen family (both Qwen3 and Qwen3.5 MoE): results show no correlation with chunk count:
  - On H200, Qwen3.5-MoE fails even without chunking (`num_chunks = 1`).
  - Failures are hardware-dependent rather than chunk-dependent: configurations that pass on H200 fail on B200, and vice versa. Running tests on Qwen3-MoE yielded identical behavior, showing no discernible pattern or trend with respect to chunk count.

Therefore, attributing convergence discrepancies in MoE models to chunk count may not be well-founded, as these architectures themselves are inherently numerically sensitive across different GPU generations.

---

#### Full benchmark data on NVIDIA B200 (BF16)

Llama-3-8B:

*% slowdown relative to 1 chunk:*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 0.00% | +50.30% | +164.60% | +396.80% | +867.10% | +1819.10% |
| 2048 | 0.00% | +43.50% | +124.60% | +297.40% | +656.30% | +1382.70% |
| 4096 | 0.00% | +22.50% | +76.10% | +180.30% | +400.90% | +857.60% |
| 8192 | 0.00% | +16.10% | +41.20% | +103.70% | +225.80% | +482.40% |
| 16384 | 0.00% | +9.10% | +21.00% | +50.70% | +115.10% | +243.70% |
| 32768 | 0.00% | +1.80% | +10.80% | +20.70% | +48.40% | +111.40% |

*% Peak Loss VRAM reduction relative to 1 chunk:*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 0.00% | 2.36% | 3.56% | 4.13% | 4.44% | 4.59% |
| 2048 | 0.00% | 4.53% | 6.77% | 7.91% | 8.47% | 8.75% |
| 4096 | 0.00% | 8.22% | 12.35% | 14.40% | 15.44% | 15.94% |
| 8192 | 0.00% | 14.01% | 21.01% | 24.52% | 26.26% | 27.15% |
| 16384 | 0.00% | 21.59% | 32.38% | 37.77% | 40.48% | 41.82% |
| 32768 | 0.00% | 21.25% | 44.38% | 51.78% | 55.47% | 57.32% |

*Full fwd+bwd runtime (ms):*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 6.2 | 9.3 | 16.4 | 30.8 | 60.0 | 119.0 |
| 2048 | 8.1 | 11.6 | 18.1 | 32.0 | 61.0 | 119.5 |
| 4096 | 12.6 | 15.4 | 22.2 | 35.3 | 63.1 | 120.6 |
| 8192 | 21.5 | 25.0 | 30.4 | 43.9 | 70.1 | 125.4 |
| 16384 | 40.6 | 44.3 | 49.2 | 61.3 | 87.4 | 139.7 |
| 32768 | 82.0 | 83.5 | 90.8 | 99.0 | 121.7 | 173.4 |

*Peak Loss VRAM (MiB):*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 5285 | 5160 | 5097 | 5066 | 5050 | 5042 |
| 2048 | 5552 | 5301 | 5176 | 5113 | 5082 | 5066 |
| 4096 | 6084 | 5584 | 5333 | 5208 | 5145 | 5114 |
| 8192 | 7150 | 6148 | 5648 | 5397 | 5272 | 5209 |
| 16384 | 9282 | 7278 | 6276 | 5776 | 5525 | 5400 |
| 32768 | 13547 | 10669 | 7535 | 6533 | 6033 | 5781 |

---

Qwen2.5-7B:

*% slowdown relative to 1 chunk:*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 0.00% | +56.10% | +173.00% | +414.80% | +904.50% | +1892.30% |
| 2048 | 0.00% | +44.60% | +130.50% | +305.80% | +676.00% | +1420.20% |
| 4096 | 0.00% | +21.30% | +70.10% | +172.20% | +382.20% | +821.30% |
| 8192 | 0.00% | +13.30% | +38.10% | +94.30% | +211.40% | +456.50% |
| 16384 | 0.00% | +10.30% | +24.50% | +56.90% | +117.30% | +251.60% |
| 32768 | 0.00% | +4.80% | +13.40% | +22.80% | +50.30% | +112.60% |

*% Peak Loss VRAM reduction relative to 1 chunk:*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 0.00% | 2.71% | 4.05% | 4.71% | 5.06% | 5.23% |
| 2048 | 0.00% | 5.08% | 7.64% | 8.91% | 9.54% | 9.87% |
| 4096 | 0.00% | 9.21% | 13.79% | 16.11% | 17.26% | 17.82% |
| 8192 | 0.00% | 15.44% | 23.15% | 27.00% | 28.94% | 29.91% |
| 16384 | 0.00% | 23.33% | 34.99% | 40.83% | 43.73% | 45.20% |
| 32768 | 0.00% | 19.84% | 47.02% | 54.85% | 58.77% | 60.72% |

*Full fwd+bwd runtime (ms):*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 6.3 | 9.8 | 17.2 | 32.5 | 63.3 | 125.6 |
| 2048 | 8.3 | 12.0 | 19.1 | 33.6 | 64.2 | 125.7 |
| 4096 | 13.7 | 16.7 | 23.4 | 37.4 | 66.2 | 126.5 |
| 8192 | 23.7 | 26.9 | 32.7 | 46.1 | 73.9 | 132.0 |
| 16384 | 41.9 | 46.2 | 52.1 | 65.8 | 91.0 | 147.3 |
| 32768 | 85.6 | 89.7 | 97.1 | 105.1 | 128.6 | 181.9 |

*Peak Loss VRAM (MiB):*

| BT \ num_chunks | 1 | 2 | 4 | 8 | 16 | 32 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 5520 | 5371 | 5296 | 5260 | 5241 | 5231 |
| 2048 | 5830 | 5534 | 5385 | 5310 | 5274 | 5255 |
| 4096 | 6452 | 5858 | 5562 | 5413 | 5338 | 5302 |
| 8192 | 7696 | 6508 | 5914 | 5618 | 5469 | 5394 |
| 16384 | 10184 | 7808 | 6620 | 6026 | 5730 | 5581 |
| 32768 | 15161 | 12153 | 8033 | 6845 | 6251 | 5955 |
